[vendor change] Add metrics command to istioctl experimental cli

douglas-reid opened the pull request:
This PR adds a new command for retrieving service-level metrics
for services within an Istio service mesh. In combination with
the `watch` command, this tool may be used to display a rudimentary
service dashboard from the commandline.

This command requires the deployment of a prometheus instance for
monitoring the mesh. It discovers a prometheus pod, establishes a
port-forward to that pod, and executes a series of queries to extract
the metrics for display.

Currently, this command pulls all metrics from the current time, 
calculating rates and latencies over a time window of 1 minute. In 
the future, it will be possible to add support for flexible time
windows.

Example usage (bookinfo example):

```
$ istioctl experimental metrics productpage reviews ratings details
productpage:
  Total RPS:     7.872870
  Error RPS:     0.000000
  P50 Latency:   40ms
  P90 Latency:   80ms
  P99 Latency:   98ms
reviews:
  Total RPS:     7.909235
  Error RPS:     0.000000
  P50 Latency:   4ms
  P90 Latency:   9ms
  P99 Latency:   21ms
ratings:
  Total RPS:     5.309187
  Error RPS:     0.000000
  P50 Latency:   2ms
  P90 Latency:   4ms
  P99 Latency:   4ms
details:
  Total RPS:     7.872870
  Error RPS:     0.000000
  P50 Latency:   3ms
  P90 Latency:   38ms
  P99 Latency:   48ms
``` 

This tool is intended primarily to aid with debugging, as discovering
what is happening with a mesh and/or a particular service can be somewhat
cumbersome.

Reviewers: please let me know if there is a more appropriate place for 
such a tool and if there is more/different information that you think
is relevant to display for a service.

Vendor PR: https://github.com/istio/vendor-istio/pull/58

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: 
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ayj**

Assign the PR to them by writing `/assign @ayj` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[istioctl/OWNERS](https://github.com/istio/istio/blob/master/istioctl/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ayj"]} -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 11278793d3318106d17679c351067993e603c26c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/4536/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 11278793d3318106d17679c351067993e603c26c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/4536/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | 11278793d3318106d17679c351067993e603c26c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-presubmit/9307/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: 
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ldemailly**

Assign the PR to them by writing `/assign @ldemailly` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ldemailly"]} -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 11278793d3318106d17679c351067993e603c26c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/4536/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | f51fbad577bc2a1398778acd36001e894adb462e | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-presubmit/9358/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | f51fbad577bc2a1398778acd36001e894adb462e | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-presubmit/9358/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | f51fbad577bc2a1398778acd36001e894adb462e | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/4588/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | f79af5e3a995ec9113c7f68fa3308c365cb138e8 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/4687/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | f79af5e3a995ec9113c7f68fa3308c365cb138e8 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-presubmit/9452/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | f79af5e3a995ec9113c7f68fa3308c365cb138e8 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-presubmit/9452/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | 86ac77d5a0eb38cde9e027944f484ed6506ad1e3 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/4690/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 86ac77d5a0eb38cde9e027944f484ed6506ad1e3 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/4690/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | 86ac77d5a0eb38cde9e027944f484ed6506ad1e3 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-presubmit/9456/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: 
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ayj**

Assign the PR to them by writing `/assign @ayj` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[istioctl/OWNERS](https://github.com/istio/istio/blob/master/istioctl/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ayj"]} -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | ae9f4ee29ddd0c45331f219fabe6d6b17b41df67 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/5977/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | ae9f4ee29ddd0c45331f219fabe6d6b17b41df67 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-unit-tests/5977/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | ae9f4ee29ddd0c45331f219fabe6d6b17b41df67 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/istio-presubmit/9948/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: 
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ldemailly**

Assign the PR to them by writing `/assign @ldemailly` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ldemailly"]} -->

douglas-reid commented:
Reviewers, PTAL.

I updated this PR to reflect the new vendoring solution. Vendor changes are in a separate commit from the code changes. Running `make depend.update` led to a large number of deletions in vendor, i suspect from previous PRs refactoring. The only dep I have added is the port-forward stuff from client-go (two files, a doc.go and the impl).

ayj commented on istioctl/cmd/istioctl/metrics.go:
```diff
@@ -0,0 +1,270 @@
+	fw, readyCh, err := buildPortForwarder(client, config, promPod.Name, port)
+	if err != nil {
+		return fmt.Errorf("could not build port forwarder for prometheus: %v", err)
+	}
```
`defer fw.Close()` here? And ignore the error in case we exited because `fw.ForwardPorts` returned an error.

douglas-reid replied:
`fw.ForwardPorts()` calls `defer fw.Close()` as the first line in its impl. So, if there is an error with that path, `fw.Close()` will be called.

ayj replied:
Should `fw.Close()` also be called for other error patches (line 127, line 134)?

douglas-reid replied:
ok. moved the defer to cover those.

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | e0129f759ced2dd8bd690c1d8d0ad5f02834462d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/e2e-bookInfoTests-envoyv2-v1alpha3/72/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | a15c6a16498374e7d4f032826c351b291c7d24e1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/e2e-bookInfoTests-envoyv2-v1alpha3/80/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented:
@kyessenov yeah, there are a lot of places we can go with this, but I wanted to get something in place before investing a bunch of time/energy.

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | aa1ad4d234e94a9d2230790fdb86d9080c0b6a0c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/e2e-bookInfoTests-envoyv2-v1alpha3/90/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/4945#pullrequestreview-114460989" title="LGTM">ayj</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ldemailly**

Assign the PR to them by writing `/assign @ldemailly` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ldemailly"]} -->

istio-testing commented:
New changes are detected. LGTM label has been removed.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/4945#pullrequestreview-114460989" title="LGTM">ayj</a>*, *<a href="https://github.com/istio/istio/pull/4945#issuecomment-383687908" title="LGTM">ldemailly</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)~~ [ldemailly]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 1cdf009cecdf5a290fcbd30870f8053ec6a15d16 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/e2e-bookInfoTests-envoyv2-v1alpha3/147/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 1cdf009cecdf5a290fcbd30870f8053ec6a15d16 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4945/e2e-bookInfoTests-envoyv2-v1alpha3/159/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
